### ISSUE TITLE
2nd Derivative doesn't appear to work with `grad_outputs` and `autograd.grad`

### ISSUE THREAD

**Issue by JulianKnodt** [pytorch/pytorch#60027]
## 🐛 Bug

I hoped to compute higher order derivatives of a  vectorized function `(N, V) -> (N, 1)`, and found on forums to use the `autograd.grad` functionality. I tried to use this on my code, but found that for the 2nd derivative it reported that the input appeared to be unused, which seems like incorrect behavior, since the gradient of the input does have some effect on the output.

## To Reproduce

```
import torch
import torch.nn as nn
x = torch.rand(40, 12, requires_grad=True)
A = nn.Linear(12, 1)
y = A(x)
g, = torch.autograd.grad(inputs=x, outputs=y, grad_outputs=torch.ones_like(y))
g2, = torch.autograd.grad(inputs=g, outputs=y, grad_outputs=torch.ones_like(y))
# Should have panicked here, since g is unused.
```

## Expected behavior

I expected g2 to be either 0, or be a tensor of shape x.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): '1.8.1+cu102'
 - OS (e.g., Linux): Both Darwin & Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): N/A
 - Python version: Python 3.8.5
 - CUDA/cuDNN version: 11.0, present on CPU as well
 - GPU models and configuration: GTX 1060
 - Any other relevant information:
It appears to be present on both CPU and GPU, so I don't think it's related to CUDA at all.

## Additional context

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by albanD** [comment]
Hi,

You can see the doc for `autograd.grad` here: https://pytorch.org/docs/stable/generated/torch.autograd.grad.html?highlight=grad#torch.autograd.grad
You need to pass `create_graph=True` to the function if you want to be able to differentiate through the computation.

Also if you are ok with Tensor that are un-used and you prefer to get a None rather than an error, you can pass `allow_unused=True`.

```python
import torch
import torch.nn as nn
x = torch.rand(40, 12, requires_grad=True)
A = nn.Linear(12, 1)
y = A(x)
g, = torch.autograd.grad(inputs=x, outputs=y, grad_outputs=torch.ones_like(y), create_graph=True)
g2, = torch.autograd.grad(inputs=g, outputs=y, grad_outputs=torch.ones_like(y), allow_unused=True)
# Will give None for g2
```

**Comment by JulianKnodt** [comment]
Hi @albanD,
Sorry for not including that in the initial issue. That does make it so it gives `None` for the first pass through, but that doesn't appear correct, I would expect it to give 0, since the 2nd derivative of `y = ax` is 0, rather than `None` which would imply that it did not affect the output. If you change `y = A(x) --> y = A(x.square())`, this also returns `None`. I expect it to return non-None, so I guess what I'm either misunderstanding or think is an error is that the output is returning `None` as opposed to a tensor. I understand you get a lot of noise on these issues, so apologies for my persistence.

**Comment by albanD** [comment]
Ho if that's what you want to compute, then the problem is with what you give as input/output to the second grad() call.
You most likely want g as outputs and `x` (or `y`) as inputs.

**Comment by JulianKnodt** [comment]
Hm, this doesn't seem to work either, possibly due to a dimension mismatch. I tried
```
torch.autograd.grad(inputs=y, outputs=g, allow_unused=True, retain_graph=True, grad_outputs=torch.ones_like(g))
# => (None,)
and
torch.autograd.grad(inputs=x, outputs=g, allow_unused=True, retain_graph=True, grad_outputs=torch.ones_like(g))
# => (None,)
```
Just to be precise and avoid confusion, I want to compute `d/dx (d (f(x)) / dx)`

**Comment by albanD** [comment]
You did copy pasted the code above properly.
The first grad() call should have `create_graph=True`.

And if you want twice derivatives wrt to x, then both `inputs=` should be `x`.
